src/cargo/ops/registry.rs

```diff
@@ -42,10 +42,13 @@ pub struct PublishOpts<'cfg> {
 pub fn publish(ws: &Workspace, opts: &PublishOpts) -> CargoResult<()> {
     let pkg = ws.current()?;
 
-    if !pkg.publish() {
-        bail!("some crates cannot be published.\n\
-               `{}` is marked as unpublishable", pkg.name());
+    if let &Some(ref allowed_registries) = pkg.publish() {
+        if opts.registry.is_none() || !allowed_registries.contains(&opts.registry.clone().unwrap()) {
```

Comment:
Perhaps this could be:


```rust
if let Some(ref registry) = opts.registry {
    if !allowed_registries.contains(registry) {
        bail!(...);
    }
}
```

Reply:
This will not work as if there are allowed registries setup, but we have not provided --registry then we also need to block. This covers the case of pushing to crates.io when publish is set.

Reply:
Er so what I mostly mean here is getting rid of the `is_none` + `clone` + `unwrap`, basically restructuring this to avoid all the function calls (and make it a little easier to read)

Reply:
I have got rid of the is_none, unwrap etc.